power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -651,106 +710,64 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
+                return calculate_power_flow_<sym>(options.err_tol, options.max_iter);
+            }
+            assert(options.optimizer_type == OptimizerType::no_optimization);
+            if constexpr (std::derived_from<calculation_type, state_estimation_t>) {
+                return calculate_state_estimation_<sym>(options.err_tol, options.max_iter);
+            }
+            if constexpr (std::derived_from<calculation_type, short_circuit_t>) {
+                return calculate_short_circuit_<sym>(options.short_circuit_voltage_scaling);
+            }
+            throw UnreachableHit{"MainModelImpl::calculate", "Unknown calculation type"};
+        }();
+
         return optimizer::get_optimizer<MainModelState, ConstDataset>(
-                   options.optimizer_type, options.optimizer_strategy,
-                   calculate_power_flow_<sym>(options.err_tol, options.max_iter),
+                   options.optimizer_type, options.optimizer_strategy, calculator,
                    [this](ConstDataset update_data) { this->update_component<permanent_update_t>(update_data); },
                    *meta_data_)
             ->optimize(state_, options.calculation_method);
     }
 
-    // Single load flow calculation, propagating the results to result_data
-    template <symmetry_tag sym>
-    void calculate_power_flow(Options const& options, MutableDataset const& result_data, Idx pos = 0) {
+    // Single calculation, propagating the results to result_data
+    void calculate(Options options, MutableDataset const& result_data, Idx pos = 0) {
         assert(construction_complete_);
-        auto const math_output = calculate_power_flow<sym>(options);
 
-        if (pos != ignore_output) {
-            output_result(math_output, result_data, pos);
-        }
-    }
+        if (options.calculation_type == CalculationType::short_circuit) {
+            auto const faults = state_.components.template citer<Fault>();
+            auto const is_three_phase = std::ranges::all_of(
+                faults, [](Fault const& fault) { return fault.get_fault_type() == FaultType::three_phase; });
+            options.calculation_symmetry =
+                is_three_phase ? CalculationSymmetry::symmetric : CalculationSymmetry::asymmetric;
+        };
 
-    // Batch load flow calculation, propagating the results to result_data
-    template <symmetry_tag sym>
-    BatchParameter calculate_power_flow(Options const& options, MutableDataset const& result_data,
-                                        ConstDataset const& update_data) {
-        return batch_calculation_(
-            [&options](MainModelImpl& model, MutableDataset const& target_data, Idx pos) {
-                auto sub_opt = options; // copy
-                sub_opt.err_tol = pos != ignore_output ? options.err_tol : std::numeric_limits<double>::max();
-                sub_opt.max_iter = pos != ignore_output ? options.max_iter : 1;
+        calculation_type_symmetry_func_selector(
```

Comment:
I think it is worth to briefly discuss the implementation details of how the `func_selectors` are used and work within the main mode, and how they reduced and simplified a bunch of code, to be used within the `calculate` function, in a knowledge sharing session.

Reply:
before or after merge?